code/modules/reagents/grinders.dm

```diff
@@ -0,0 +1,370 @@
+	var/is_chamber_empty = 0
+	var/is_beaker_ready = 0
+	var/processing_chamber = ""
+	var/beaker_contents = ""
+	var/dat = list()
+
+	if(!inuse)
+		for (var/obj/item/O in holdingitems)
+			processing_chamber += "\A [O.name]<BR>"
+
+		if (!processing_chamber)
+			is_chamber_empty = 1
+			processing_chamber = "Nothing."
+		if (!beaker)
+			beaker_contents = "<B>No beaker attached.</B><br>"
+		else
+			is_beaker_ready = 1
+			beaker_contents = "<B>The beaker contains:</B><br>"
+			var/anything = 0
+			for(var/datum/reagent/R in beaker.reagents.reagent_list)
+				anything = 1
+				beaker_contents += "[R.volume] - [R.name]<br>"
+			if(!anything)
+				beaker_contents += "Nothing<br>"
+
+
+		dat += {"
+	<b>Processing chamber contains:</b><br>
+	[processing_chamber]<br>
+	[beaker_contents]<hr>
+	"}
+		if (is_beaker_ready && !is_chamber_empty && !(stat & (NOPOWER|BROKEN)))
+
+			// AUTOFIXED BY fix_string_idiocy.py
+			// C:\Users\Rob\\documents\\\projects\vgstation13\code\\modules\reagents\Chemistry-Machinery.dm:1016: dat += "<A href='?src=\ref[src];action=grind'>Grind the reagents</a><BR>"
+			dat += {"<A href='?src=\ref[src];action=process'>Process contents</a><BR>
+				<A href='?src=\ref[src];action=extract'>Extract reagents</a><BR><BR>"}
+			// END AUTOFIX
+		if(holdingitems && holdingitems.len > 0)
+			dat += "<A href='?src=\ref[src];action=eject'>Eject contents</a><BR>"
+		if (beaker)
+			dat += "<A href='?src=\ref[src];action=detach'>Detach beaker</a><BR>"
+	else
+		dat += "Please wait..."
+	dat = list2text(dat)
+	var/datum/browser/popup = new(user, "reagentgrinder", "All-In-One Grinder", src)
+	popup.set_content(dat)
+	popup.open()
+	onclose(user, "reagentgrinder")
+	return
+
+
+/obj/machinery/reagentgrinder/Topic(href, href_list)
+	if(..())
+		return
+	usr.set_machine(src)
+	switch(href_list["action"])
+		if ("process")
+			grind(1)
+		if("extract")
+			grind(0)
+		if("eject")
+			eject()
+		if ("detach")
+			detach()
+	src.updateUsrDialog()
+	return
+
+/obj/machinery/reagentgrinder/proc/detach()
+
+
+	if (usr.stat != 0)
+		return
+	if (!beaker)
+		return
+	beaker.loc = src.loc
+	if(istype(beaker, /obj/item/weapon/reagent_containers/glass/beaker/large/cyborg))
+		var/mob/living/silicon/robot/R = beaker:holder:loc
+		if(R.module_state_1 == beaker || R.module_state_2 == beaker || R.module_state_3 == beaker)
+			beaker.loc = R
+		else
+			beaker.loc = beaker:holder
+	beaker = null
+	update_icon()
+
+/obj/machinery/reagentgrinder/proc/eject()
+
+
+	if (usr.stat != 0)
+		return
+	if (holdingitems && holdingitems.len == 0)
+		return
+
+	for(var/obj/item/O in holdingitems)
+		O.loc = src.loc
+		holdingitems -= O
+	holdingitems = list()
+
+/obj/machinery/reagentgrinder/proc/is_allowed(var/obj/item/I)
+	if(I.grindable_reagent || istype(I, /obj/item/weapon/reagent_containers/food/snacks) || istype(I,/obj/item/weapon/grown)) return 1
+	return 0
+
+/obj/machinery/reagentgrinder/proc/grind(var/process = 0) //passed to grind_item
+	power_change()
```

Comment:
Is this (and the 2 lines below) even necessary?

It only gets called from `Topic()` which is already doing checking for power and such.